### PR TITLE
Added Pipelines.

### PR DESCRIPTION
**Pull request by BuzzHari**
I tried to leave majority of the code as you have written it.
I did add an another structure to the command.h to handle multiple commands. I have also commented the code where I thought it was needed. 
Multiple pipes can be executed, but the parsing is really naive, I felt not to complicate it, because a proper and easier parser can be written using Bison/Yacc.

### CODE REVIEW

File: src/command.c [hunk 1]
```diff
@@ -2,15 +2,28 @@
...
+    for(size_t i = 0; i <= command_table->no_of_commands; i++){
+        for (size_t j = 0; j < command_table->cmd[i].size; j++) {
+            free(command_table->cmd[i].argv[j]);
+        }
+        free(command_table->cmd[i].argv);
+        command_table->cmd[i].size = 0;
     }
-    free(command->argv);
-    command->size = 0;
-}
```
Comment (xorz57) [inline src/command.c]: Don't modify the `command_destructor` function. Make a `command_table_destructor` function.

File: src/command.c [hunk 2]
```diff
@@ -2,15 +2,28 @@
...
+
+    //allocating table for one command initally.
+    command_table->cmd = (command*) malloc(sizeof(command)); 
+    if(!command_table->cmd){
+        free(command_table->cmd);
+        exit(EXIT_FAILURE);
+    }
+    command_table->cmd->argv = NULL;
+    command_table->cmd->size = 0;
 }
```
Comment (xorz57) [inline src/command.c]: Don't modify the `command_constructor` function. Make a `command_table_constructor` function.

File: src/main.c [hunk 3]
```diff
@@ -2,23 +2,23 @@
 #include "parse.h"
 #include "prompt.h"
 #include "read.h"
-
+#include <stdio.h>
```
Comment (xorz57) [inline src/main.c]: We don't need this header here.

File: src/parse.c [hunk 4]
```diff
@@ -4,46 +4,67 @@
...
+        if(*token == '|'){
+            command_table.no_of_commands += 1;
+            cmd_index += 1;
+            command_table.cmd = realloc(command_table.cmd, sizeof(command_t)*(command_table.no_of_commands+1));
+            if(!command_table.cmd){
+                free(command_table.cmd);
+                exit(EXIT_FAILURE);
+            }
+            command_table.cmd[cmd_index].argv = NULL;
+            command_table.cmd[cmd_index].size = 0;
```
Comment (xorz57) [inline src/parse.c]: You could use `command_constructor` here if you didn't delete it.

File: src/command.h [hunk 5]
```diff
@@ -3,13 +3,19 @@
...
     size_t size;
-} command_t;
+} command;
 
-void command_constructor(command_t * command);
+//Command table.
+typedef struct command_t{
+    int no_of_commands;
+    struct command *cmd;
+}command_t;
```
Comment (xorz57) [inline src/command.h]: Name this `command_table_t`. Rename `no_of_commands` to `size` and `cmd` to `command`.

File: src/command.h [hunk 6]
```diff
@@ -3,13 +3,19 @@
 
 #include <stddef.h>
 
-typedef struct command_t {
+//Single Command.
+typedef struct command {
     char ** argv;
     size_t size;
-} command_t;
+} command;
```
Comment (xorz57) [inline src/command.h]: Keep the original name.

File: src/parse.c [hunk 7]
```diff
@@ -4,46 +4,67 @@
...
+    //To keep track of number of commands.
+    //As it's 0 indexed. cmd_index = 0, indicates one command.
+    size_t cmd_index = 0; 
 
     for (i = 0, str = buffer.str; ; ++i, str = NULL) {
         char * token = strtok_r(str, del, &saveptr);
-        if (!token) break;
-        command.size++;
+        if (!token) 
+            break;
```
Comment (xorz57) [inline src/parse.c]: Make it one line.

File: src/parse.c [hunk 8]
```diff
@@ -4,46 +4,67 @@
...
+        
+        //Checking for PIPE('|'), If pipe, allocating space for the new command.
+        if(*token == '|'){
+            command_table.no_of_commands += 1;
+            cmd_index += 1;
+            command_table.cmd = realloc(command_table.cmd, sizeof(command_t)*(command_table.no_of_commands+1));
+            if(!command_table.cmd){
+                free(command_table.cmd);
```
Comment (xorz57) [inline src/parse.c]: You could use `command_destructor` here.

**Comment by xorz57** [comment]
The shell works! But the code needs some changes.

**Comment by BuzzHari** [comment]
I have made the requested changes. Also I caught a mistake I made, I never appended "NULL" to the commands before the pipe, I am surprised it didn't give any unwanted outputs or maybe I didn't test it enough.
If there's any more changes you would like, do tell.

File: src/execute.h [hunk 9]
```diff
@@ -3,6 +3,7 @@
 
 #include "command.h"
 
-int execute(const command_t command);
+//Pointer to the command array.
+int execute(const command_table_t command);
```
[the remark this answers is not in the document]
Comment (BuzzHari) [reply src/execute.h]: I should have taken care of this. My bad.

File: src/main.c [hunk 10]
```diff
@@ -7,25 +7,26 @@
...
-    command_t command;
+    //This is basically a Command Table now.
+    command_table_t command_table;
 
     while (1) {
         prompt("shell");
 
         buffer_constructor(&buffer);
-        command_constructor(&command);
+        //command_table_constructor(&command_table);
```
[the remark this answers is not in the document]
Comment (BuzzHari) [reply src/main.c]: Oh okay.

**Pull request merged**